File: Insert_Delete_GetRandom_O(1).py
class RandomizedSet:

    def __init__(self):
        self.values = []
        self.valuesIdx = {} # value: index

    def insert(self, val: int) -> bool:
        if val in self.valuesIdx:
            return False
        
        self.valuesIdx[val] = len(self.values)
        self.values.append(val)

        return True

    def remove(self, val: int) -> bool:
        if val not in self.valuesIdx:
            return False
        
        index = self.valuesIdx[val]
        self.valuesIdx[self.values[-1]] = index
        del self.valuesIdx[val]
        self.values[index] = self.values[-1]
        self.values.pop()

        return True

    def getRandom(self) -> int:
        index = random.randint(0, len(self.values) - 1)
        return self.values[index]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: Insert_Delete_GetRandom_O(1).java
class RandomizedSet {

    private List<Integer> values;
    private Map<Integer, Integer> valuesIdx;

    public RandomizedSet() {
        this.values = new ArrayList<>();
        this.valuesIdx = new HashMap<>();
    }

    public boolean insert(int val) {
        if (valuesIdx.containsKey(val)) {
            return false;
        }

        valuesIdx.put(val, values.size());
        values.add(val);

        return true;
    }

    public boolean remove(int val) {
        if (!valuesIdx.containsKey(val)) {
            return false;
        }

        int index = valuesIdx.get(val);
        valuesIdx.put(values.get(values.size() - 1), index);
        valuesIdx.remove(val);
        values.set(index, values.get(values.size() - 1));
        values.remove(values.size() - 1);

        return true;
    }

    public int getRandom() {
        int index = (int) (Math.random() * values.size());
        return values.get(index);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
